Use human readable features in configuration

araspitzu opened the pull request:
Allow users to specify the desired features in human readable way, instead of the byte vector. When parsing from config we ignore unknown features but we crash if there is a known feature with unknown support, we also pretty print the features when using `/getinfo` API or when parsing a BOLT11 invoice.

Fixes #1359

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/Features.scala:
```diff
@@ -154,4 +161,67 @@ object Features {
    */
   def areSupported(features: ByteVector): Boolean = areSupported(features.bits)
 
+  object Resolution {
```
I think the functions/contracts in this object are a bit confusing.
What do you think of the following alternative proposal:

```scala
object Resolution {
    
    private val knownFeatures: Seq[Feature] = Seq(
      OptionDataLossProtect,
      InitialRoutingSync,
      ChannelRangeQueries,
      VariableLengthOnion,
      ChannelRangeQueriesExtended,
      PaymentSecret,
      BasicMultiPartPayment,
      Wumbo,
      TrampolinePayment
    )

    case class ActivatedFeature(feature: Feature, support: FeatureSupport)

    def fromConfiguration(config: Config): Seq[ActivatedFeature] = ???
    
    def toByteVector(features: Seq[ActivatedFeature]): ByteVector = ???

    def fromByteVector(features: ByteVector): Seq[ActivatedFeature] = ???

}
```

The `ActivatedFeature` would replace the `HumanReadableFeature` in `Eclair.scala`.
We could maybe even create `case class ActivatedFeatures(features: ActivatedFeature*)` and put `fromByteVector/toByteVector` as methods on this class? I think that would make the contracts clearer.

araspitzu replied:
Currently the role of `HumanReadableFeature` is to help render the features as JSON, notice how the fields `name` and `support` end up serialized as JSON and that's what i wanted to achieve, it makes sense for anyone consuming the API to have that kind of object in the response.

I like the suggestion of introducing `ActivatedFeature` but i think it's even better if we take it one step further and incapsulate them inside a `Features` object, this would be carried around in the codebase instead of the ByteVector features.

```
case class Features(features: Seq[ActivatedFeature]){
  def hasFeature(f: Feature): Boolean
  def toByteVector: ByteVector
  def apply(raw: ByteVector): Features
}
```
WDYT?

t-bast replied:
> how the fields name and support end up serialized as JSON

But that's true as well with `ActivatedFeature`, right? The only difference is that the JSON is `{ feature: data_loss_protect, support: mandatory }` instead of `name`?

> I like the suggestion of introducing ActivatedFeature but i think it's even better if we take it one step further and incapsulate them inside a Features object, this would be carried around in the codebase instead of the ByteVector features.

I really like that! That means `NodeParams` would directly contain this `Features` object instead of a `ByteVector`, it would be really nice.

pm47 replied:
Let's see what the `ActivatedFeature` idea lead us to, but imo this `Resolution` class is a *lot* of code for what it's doing.

araspitzu commented:
> This is going to be a nice improvement!
> I'd like to agree on a clearer interface/functions before reviewing the tests, see my comment below.
> I think you're also missing config migration (in `makeNodeParams`) and corresponding tests?

Gotcha, i added the config migration + test in the latest commit

pm47 commented on eclair-core/src/main/scala/fr/acinq/eclair/Features.scala:
```diff
@@ -154,4 +161,67 @@ object Features {
   def areSupported(features: ByteVector): Boolean = areSupported(features.bits)
 
+  object Resolution {
+    private val knownFeatures: Seq[Feature] = Seq(
```
How about using some reflection to guard us against forgetting new future features? How ugly would it look like?

araspitzu replied:
Not too ugly but we'd have to add a new dependency for `scala-reflect`:
```scala
    import scala.reflect.runtime.universe._
    import scala.reflect.runtime.{ universe => runtime }
    val mirror = runtime.runtimeMirror(ClassLoader.getSystemClassLoader)
    val subclasses = typeOf[Feature].typeSymbol.asClass.knownDirectSubclasses
    lazy val knownFeatures = subclasses.map({ desc =>
      val mod = mirror.staticModule(desc.asClass.fullName)
      mirror.reflectModule(mod).instance
    })
```

araspitzu replied:
As a compromise i added the test to guard us against forgetting new features but without using reflection at runtime. b2441ed

pm47 commented on eclair-core/src/main/scala/fr/acinq/eclair/Features.scala:
```diff
@@ -154,4 +161,67 @@ object Features {
+      if(!config.hasPath(s"features.$name")) None
+      else {
+        config.getString(s"features.$name") match {
+          case "mandatory" => Some(Mandatory)
+          case "optional" => Some(Optional)
```
A bit over the top?
```suggestion
          case support if support == Mandatory.toString => Some(Mandatory)
          case support if support == Optional.toString => Some(Optional)
```

araspitzu commented:
Latest commit 1dd4ba5 introduces `ActivatedFeature` and `InactiveFeature`, the former is used to store the known features and the latter those that we don't know. It's necessary to store the unknown features as well because we want to validate the remote peer's features and doing something like:
```
 case class Features(...) {
    require(noUnknownEvenFeatures)
 }
```
seems quite ugly because it would fail when parsing the `Init` from remote and it's preferable to use the existing validation methods. 

I addressed most of the feedback in the last commit including the json serialization (now `Features` has its own serializer) and refactored the `Resolution` object into something smaller inside `Features`. 

I also took the chance to rename the invoice features to `PaymentRequestFeatures` as the naming caused a bit of confusion. Note that i added the `Features` inside `wire.Init` but not in all the other wire messages, let me know if we want to propagate this down to all the other messages (even though there isn't a big advantage in doing so).

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala:
```diff
@@ -147,6 +147,10 @@ object NodeParams {
       case (old, new_) => require(!config.hasPath(old), s"configuration key '$old' has been replaced by '$new_'")
     }
 
+    // since v0.3.5 features cannot be a byte vector (hex string)
```
```suggestion
    // since v0.4.1 features cannot be a byte vector (hex string)
```

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/NodeParams.scala:
```diff
@@ -147,6 +147,10 @@ object NodeParams {
     }
 
+    // since v0.3.5 features cannot be a byte vector (hex string)
+    val isFeatureByteVector = config.getValue("features").valueType() == ConfigValueType.STRING
+    require(!isFeatureByteVector, "configuration key 'features' have moved from bytevector to human readable (ex: 'feature-name' = optional/mandatory)")
```
nit:

```suggestion
    require(config.getValue("features").valueType() != ConfigValueType.STRING, "configuration key 'features' have moved from bytevector to human readable (ex: 'feature-name' = optional/mandatory)")
```

araspitzu replied:
Not sure about this, isn't it too big to be inline?

t-bast replied:
It's only the text that is long, not really the equality check, so it doesn't shock me. But if you prefer the current version don't hesitate to ignore!

t-bast commented on eclair-core/src/test/scala/fr/acinq/eclair/io/PeerConnectionSpec.scala:
```diff
@@ -191,19 +193,19 @@ class PeerConnectionSpec extends TestKitBaseClass with FixtureAnyFunSuiteLike wi
   test("don't sync if whitelist doesn't contain peer", Tag("sync-whitelist-random")) { f =>
     import f._
-    val remoteInit = wire.Init(bin"0000001010000000".bytes) // bob supports channel range queries
+    val remoteInit = wire.Init(Features(bin"0000001010000000")) // bob supports channel range queries
```
I think it's not this test's responsibility to test the feature parsing from bits, better make it more readable:

```suggestion
    val remoteInit = wire.Init(Features(Set(ActivatedFeature(ChannelRangeQueries, Optional))))
```

t-bast commented on eclair-core/src/test/scala/fr/acinq/eclair/payment/MultiPartHandlerSpec.scala:
```diff
@@ -43,9 +46,19 @@ import scala.concurrent.duration._
 
 class MultiPartHandlerSpec extends TestKitBaseClass with FixtureAnyFunSuiteLike {
 
+  val features = Features(Set(
```
nit:

```suggestion
  val featuresWithMpp = Features(Set(
```

t-bast commented on eclair-core/src/test/scala/fr/acinq/eclair/payment/PaymentInitiatorSpec.scala:
```diff
@@ -50,8 +51,17 @@ class PaymentInitiatorSpec extends TestKitBaseClass with FixtureAnyFunSuiteLike
+    ActivatedFeature(ChannelRangeQueriesExtended, Optional),
+    ActivatedFeature(VariableLengthOnion, Optional)))
+
+  val featuresWithMpp = Features(defaultTestFeatures.activated + ActivatedFeature(BasicMultiPartPayment, Optional))
```
```suggestion
  val featuresWithMpp = Features(defaultTestFeatures.activated + ActivatedFeature(PaymentSecret, Optional) + ActivatedFeature(BasicMultiPartPayment, Optional))
```

t-bast commented on eclair-core/src/test/scala/fr/acinq/eclair/payment/PaymentPacketSpec.scala:
```diff
@@ -47,6 +48,8 @@ class PaymentPacketSpec extends AnyFunSuite with BeforeAndAfterAll {
 
   implicit val log: akka.event.LoggingAdapter = akka.event.NoLogging
 
+  val variableLengthOnionFeature = Features(Set(ActivatedFeature(VariableLengthOnion, Optional)))
```
This should be in the object below, it's currently duplicated (see line 385)

t-bast commented on eclair-core/src/main/scala/fr/acinq/eclair/io/PeerConnection.scala:
```diff
@@ -105,7 +106,7 @@ class PeerConnection(nodeParams: NodeParams, switchboard: ActorRef, router: Acto
           // We use a long enough mask to account for future features.
           // TODO: remove that once eclair-mobile is patched.
-          val tweakedFeatures = BitVector.bits(nodeParams.features.bits.reverse.toIndexedSeq.zipWithIndex.map {
+          val tweakedFeatures = BitVector.bits(nodeParams.features.toByteVector.bits.reverse.toIndexedSeq.zipWithIndex.map {
```
You can now make this code a lot more readable. Instead of modifying a `ByteVector`, just modify the `Features` object to remove the `basic_mpp` and `payment_secret` features from the set of activated features.

araspitzu replied:
Yes, this is done in 962d2a5 and it's been extracted into a function.

araspitzu commented on eclair-core/src/main/scala/fr/acinq/eclair/Features.scala:
```diff
@@ -38,11 +40,87 @@ sealed trait Feature {
 
+case class ActivatedFeature(feature: Feature, support: FeatureSupport)
+
+case class InactiveFeature(bitIndex: Int)
```
Went for `UnknownFeature` in 8840426

araspitzu commented on eclair-core/src/main/scala/fr/acinq/eclair/router/Announcements.scala:
```diff
@@ -73,17 +73,17 @@ object Announcements {
       rgbColor = color,
       alias = alias,
-      features = features,
+      features = features.toByteVector,
```
I wasn't sure if we wanted to type all of them so i started with the `Init`, with the latest changes (eab0654) i typed the remaining `ChannelAnnouncement` and `NodeAnnouncement`
